### PR TITLE
[PREVIEW] SSCS-3570 save answer

### PR DESCRIPTION
**Pull request by bendiggle**
- Add save Answer functionality.
- When you click save, the controller determines whether or not it has been saved or submitted and then validates the field.
- If field empty or over 10,000 characters then we re-render the page with the error messages.
- If field valid then we post the answer to the api with the draft state.
- Added tests

### CODE REVIEW

File: config/default.json [hunk 1]
```diff
@@ -2,6 +2,7 @@
   "node": {
     "port": "3000"
   },
+  "testUrl": "http://localhost:3000",
```
Comment (davet1985) [inline config/default.json]: Is this needed?

File: app/views/task-list.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{% extends "layout.html" %}
+
+{% block content %}
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+
+        <h1 class="govuk-heading-l question-header">{{ i18n.taskList.header }}</h1>
```
Comment (davet1985) [inline app/views/task-list.html]: remove the `question-header` class here

File: test/browser/question.test.js [hunk 3]
```diff
@@ -46,4 +52,23 @@ describe('Question page', () => {
...
+      .contain(i18n.question.textareaField.error.empty);
+  });
+
+  it('displays an error message above the field when you try to save an empty answer', async() => {
+    await questionPage.saveAnswer('');
+    expect(await questionPage.getElementText('#question-field-error'))
+      .contain(i18n.question.textareaField.error.empty);
+  });
+
+  if (testingLocalhost) {
```
Comment (davet1985) [inline test/browser/question.test.js]: I'm guessing this a temporary measure until the backend work is done? I'd rather avoid writing tests that can't be run against preview and AAT.

File: test/mock/services/question.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
...
+/* eslint-disable max-len, no-confusing-arrow */
 module.exports = {
   path: '/continuous-online-hearings/:onlineHearingId/questions/:questionId',
+  method: 'GET',
   template: {
     question_id: params => `${params.questionId}`,
     question_header_text: 'How do you interact with people?',
-    question_body_text: 'You said you avoid interacting with people if possible. We\'d like to know more about the times when you see friends and family.\n\nTell us about three separate occasions in 2017 that you have met with friends and family.\n\nTell us:\n\n- who you met\n\n- when\n\n- where\n\n- how it made you feel'
+    question_body_text: 'You said you avoid interacting with people if possible. We\'d like to know more about the times when you see friends and family.\n\nTell us about three separate occasions in 2017 that you have met with friends and family.\n\nTell us:\n\n- who you met\n\n- when\n\n- where\n\n- how it made you feel',
+    question_answer_text: params => params.questionId === '1' ? 'My amazing answer' : ''
```
Comment (davet1985) [inline test/mock/services/question.js]: Is this just used for testing what it looks like when an answer comes back? Just need to be not to add anything here that we can't replicate in the real service.

File: app/controllers/question.js [hunk 5]
```diff
@@ -1,33 +1,71 @@
...
+        questionId,
+        header: response.question_header_text,
+        body: response.question_body_text,
+        answer: {
+          value: response.question_answer_text
         }
-      });
+      };
+      req.session.question = question;
+      req.session.save();
```
Comment (davet1985) [inline app/controllers/question.js]: I was just trying to work out if this is required, we have `resave` set to true in the session store config.

File: app/middleware/session.js [hunk 6]
```diff
@@ -1,7 +1,8 @@
 const session = require('express-session');
 const config = require('config');
 
-const isDevelopment = process.env.NODE_ENV === 'development';
+const testUrl = config.get('testUrl');
+const isDevelopment = process.env.NODE_ENV === 'development' || testUrl.indexOf('localhost') !== -1;
```
[the remark this answers is not in the document]
Comment (davet1985) [reply app/middleware/session.js]: As we have discussed, will move this to a separate config item used specifically to drive whether or not to use secure cookies, controlled via the terraform vars.

[the remark this answers is not in the document]
Comment (davet1985) [reply app/middleware/session.js]: Happy for this ^^^ to be done separately.

File: app/views/task-list.html (hunk 2 above)
Comment (bendiggle) [reply app/views/task-list.html]: You sure? You put it there

File: test/browser/question.test.js (hunk 3 above)
Comment (bendiggle) [reply test/browser/question.test.js]: Yeah this is just until the api is ready

File: app/controllers/question.js (hunk 5 above)
Comment (bendiggle) [reply app/controllers/question.js]: Yep you're right

File: app/views/task-list.html (hunk 2 above)
Comment (davet1985) [reply app/views/task-list.html]: Only for the question header h1, this is a task list header

Comment (bendiggle) [reply app/views/task-list.html]: Oh wait thought it was the other template

Comment (davet1985) [reply app/views/task-list.html]: Looks like a copy and paste error 😱

File: test/mock/services/question.js [hunk 7]
```diff
@@ -1,9 +1,11 @@
-/* eslint-disable max-len */
+/* eslint-disable max-len, no-confusing-arrow */
```
Comment (davet1985) [inline test/mock/services/question.js]: Doesn't look like there are any more confusing arrows?

**Pull request merged**